updater/lib/dependabot/updater/group_dependency_file_batch.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# This class is responsible for tracking changes to the original files as we
+# make changes
+module Dependabot
+  class Updater
+    class GroupDependencyFileBatch
+      def initialize(initial_dependency_files)
+        @dependency_file_batch = initial_dependency_files.each_with_object({}) do |file, hash|
+          hash[file.path] = { file: file, changed: false, changes: 0 }
+        end
+
+        Dependabot.logger.debug("Starting with '#{@dependency_file_batch.count}' dependency files:")
+        debug_current_state
+      end
+
+      # Returns an array of DependencyFile objects for the current state
+      def dependency_files
+        @dependency_file_batch.map do |_path, data|
+          data[:file]
+        end
+      end
+
+      # Returns an array of DependencyFile objects that have changed at least once
+      def updated_files
+        @dependency_file_batch.filter_map do |_path, data|
+          data[:file] if data[:changed]
+        end
+      end
+
+      # Replaces the existing files
+      def merge(updated_files)
+        updated_files.each do |updated_file|
+          existing_file = @dependency_file_batch[updated_file.path]
+
+          change_count = if existing_file
+                           existing_file.fetch(:change_count, 0)
+                         else
+                           # Let's warn about this in debug mode, but otherwise tolerate this.
```

Comment:
Could updates to vendored dependencies trigger this or do they get handled somewhere else?

Reply:
I think the adapters do pass them back so this will warn on vendored files - I will make sure this is the case as it will be worth calling out a new file being injected into the set in debug mode.